### PR TITLE
Download blobs from the node that gave us the certificate.

### PR DESCRIPTION
**Pull request by afck**
## Motivation

In `receive_certificates_for_blobs` we simultaneously ask all validators for a certificate that (re-)certifies a blob. We then take the first response we get and call `receive_certificate`. That will fail because we don't have the blob yet, which we in turn request from all validators _sequentially_! Due to https://github.com/linera-io/linera-protocol/issues/3033, if the first validator we ask doesn't have it, we keep retrying a few times before we move on to the next one, which can take a long time.

## Proposal

Download the blob in question from the validator that sent us the certificate and is therefore expected to have it.

## Test Plan

For me this reduced the `process-inbox` in https://github.com/linera-io/linera-protocol/issues/3029 from 53 to 7 seconds.

## Release Plan

- These changes should be backported to the latest `main` branch.
- These changes should be backported to the latest `devnet` branch, then
    - be released in a new SDK.

## Links

- Fixes https://github.com/linera-io/linera-protocol/issues/3029.
- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-core/src/client/mod.rs [hunk 1]
```diff
@@ -1807,15 +1816,26 @@ where
         for blob_id in blob_ids {
             let mut certificate_stream = validators
                 .iter()
-                .map(|remote_node| remote_node.download_certificate_for_blob(blob_id))
+                .map(|remote_node| async move {
+                    let cert = remote_node.download_certificate_for_blob(blob_id).await?;
```
Comment (deuszx) [inline linera-core/src/client/mod.rs:1820]: Does this still keep the old semantics of running all futures in parallel thanks to `async move`?

**Review by deuszx — APPROVED** [review]

File: linera-core/src/client/mod.rs (hunk 1 above)
Comment (afck) [reply linera-core/src/client/mod.rs:1820]: Yes, it maps each `remote_node` to a future (the `async move` block) and then creates the `FuturesUnordered` from them.

File: linera-core/src/client/mod.rs [hunk 2]
```diff
@@ -1207,6 +1208,7 @@ where
         &self,
         certificate: Certificate,
         mode: ReceiveCertificateMode,
+        nodes: Option<Vec<RemoteNode<P::Node>>>,
```
Comment (ma2bd) [inline linera-core/src/client/mod.rs:1211]: > Download the blob in question from the validator that sent us the certificate and is therefore expected to have it.

This code looks like it does the opposite, i.e. downloading the certificate from the validator that has the blob. What am I missing?

Comment (afck) [reply linera-core/src/client/mod.rs:1211]: The certificate itself—the one we downloaded _because_ of the blob in question—is passed into this function. But the function also automatically downloads all the ancestors and blobs, if needed.

It currently downloads those from _all_ validators, sequentially. With this change, if a node is provided explicitly, it downloads them only from that node.

Comment (afck) [reply linera-core/src/client/mod.rs:1211]: Yes, exactly, that's a more general and far-reaching discussion we should have at some point: The right tradeoff between latency (ask all validators everything at the same time, take the first response) and bandwidth (ask them in sequence).

But at least it triggers exactly that bottleneck way less frequently that before: Before, we were simply asking the _wrong_ validator often, and then, due to #3033, waiting a long time for it.

Comment (ma2bd) [reply linera-core/src/client/mod.rs:1211]: re: tradeoff: We don't need to be faster than the fastest validator.

**Pull request merged**